frame/mosaic/src/lib.rs

```diff
@@ -0,0 +1,676 @@
+// TODO
+// 1. TEST!
+// 2. RPCs for relayer convenience.
+// 3. Refactor core logic to traits.
+// 4. Benchmarks and Weights!
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+mod decay;
+mod relayer;
+
+pub use decay::{BudgetPenaltyDecayer, Decayer};
+pub use pallet::*;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+#[frame_support::pallet]
+pub mod pallet {
+
+	use crate::{
+		decay::Decayer,
+		relayer::{RelayerConfig, StaleRelayer},
+	};
+	use codec::FullCodec;
+	use composable_traits::math::SafeArithmetic;
+	use frame_support::{
+		dispatch::DispatchResultWithPostInfo,
+		pallet_prelude::*,
+		traits::fungibles::{Inspect, Mutate, Transfer},
+		transactional, PalletId,
+	};
+	use frame_system::pallet_prelude::*;
+	use num_traits::{CheckedSub, Zero};
+	use scale_info::TypeInfo;
+	use sp_core::H256;
+	use sp_runtime::{
+		traits::{AccountIdConversion, Keccak256, Saturating},
+		DispatchError,
+	};
+	use sp_std::{fmt::Debug, str};
+
+	type AccountIdOf<T> = <T as frame_system::Config>::AccountId;
+	type BalanceOf<T> = <<T as Config>::Assets as Inspect<AccountIdOf<T>>>::Balance;
+	type BlockNumberOf<T> = <T as frame_system::Config>::BlockNumber;
+	type AssetIdOf<T> = <<T as Config>::Assets as Inspect<AccountIdOf<T>>>::AssetId;
+	type NetworkIdOf<T> = <T as Config>::NetworkId;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		type PalletId: Get<PalletId>;
+		type Assets: Mutate<AccountIdOf<Self>> + Transfer<AccountIdOf<Self>>;
+
+		type MinimumTTL: Get<BlockNumberOf<Self>>;
+		type MinimumTimeLockPeriod: Get<BlockNumberOf<Self>>;
+
+		type BudgetPenaltyDecayer: Decayer<BalanceOf<Self>, BlockNumberOf<Self>>
+			+ Clone
+			+ Encode
+			+ Decode
+			+ Debug
+			+ TypeInfo
+			+ PartialEq;
+
+		type NetworkId: FullCodec + TypeInfo + Clone + Debug + PartialEq;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::storage]
+	pub type Relayer<T: Config> =
+		StorageValue<_, StaleRelayer<T::AccountId, T::BlockNumber>, ValueQuery>;
+
+	#[derive(Clone, Debug, Encode, Decode, TypeInfo, PartialEq)]
+	pub struct AssetInfo<BlockNumber, Balance, Decayer> {
+		pub last_mint_block: BlockNumber,
+		pub budget: Balance,
+		pub penalty: Balance,
+		pub penalty_decayer: Decayer,
+	}
+
+	pub enum TransactionType {
+		Incoming,
+		Outgoing,
+	}
+
+	/// User incoming/outgoing accounts, that hold the funds for transactions to happen.
+	pub struct SubAccount<T: Config> {
+		transaction_type: TransactionType,
+		account_id: AccountIdOf<T>,
+	}
+
+	impl<T: Config> SubAccount<T> {
+		pub fn to_id(&self) -> impl Encode {
```

Comment:
consider impl AccountIdConversion

Reply:
The trait has many methods we don't want to use, would be overkill to implement them and make the feature more obscure